refactor!: Async Implementation

danwritecode opened the pull request:
Fixes #12 

Quickly forked things and created an async module that is basically a copy of the v1 module but with the api client using reqwest. I updated all tests and have most of them passing. There are a few failing due to unique constraint errors in chroma?

This is maybe a naive implementation as this requires maintaining two separate modules which have 95% of the same code but I'm also not sure of a more eloquent solution. 

Let me know and I'd be happy to implement any further changes.

Anush008 commented:
Hi @danwritecode.
I wrote the sync implementation prioritizing simplicity and avoiding the requirement of an async runtime. Which I don't feel strongly about anymore.

Let us delete the sync implementation as most real-world applications would need asynchronous behaviour. I am fine with doing a major(BREAKING) release. What do you think?

Also, please add yourself to the authors list in the `Cargo.toml`.

danwritecode commented:
@Anush008 this makes sense to me. If someone wants a basic sync implementation then can use an older version.

What do you want to call the async module? Leave it as async or call it v2?

Anush008 commented:
v2++

Thank you.

danwritecode commented:
Renamed to v2, updated all tests and docs. A couple things to note: 

1. The below test is still failing. 
```
test v2::collection::tests::test_get_from_collection ... FAILED
```
Upon inspection, i was able to get it to pass by commenting out the entire json object here: 
```rust
    pub async fn get(&self, get_options: GetOptions) -> Result<GetResult> {
        let GetOptions {
            ids,
            where_metadata,
            limit,
            offset,
            where_document,
            include,
        } = get_options;
        let mut json_body = json!({
            "ids": ids,
            "where": where_metadata,
            "limit": limit,
            "offset": offset,
            "where_document": where_document,
            "include": include
        });

        json_body
            .as_object_mut()
            .unwrap()
            .retain(|_, v| !v.is_null());

        let path = format!("/collections/{}/get", self.id);
        let response = self.api.post(&path, Some(json_body)).await?;
        let get_result = response.json::<GetResult>().await?;
        Ok(get_result)
    }
```

It seems like the chroma docs are outdated regarding the behavior of queries. `If no ids are supplied, it will return all items in the collection that match the where and where_document filters.` no longer seems to be the case, but I'm not 100% sure.

I left the test failing for now, we can discuss best way to handle. 

2. The update call no longer returns anything from chroma, i confirmed this by looking at `response.text()` from reqwest. Given this, I updated it to be a Result<()>, and then return an error if the status code from the server is not Ok.

```rust
    pub async fn update<'a>(
        &self,
        collection_entries: CollectionEntries<'a>,
        embedding_function: Option<Box<dyn EmbeddingFunction>>,
    ) -> Result<()> {
        let collection_entries = validate(false, collection_entries, embedding_function)?;

        let CollectionEntries {
            ids,
            embeddings,
            metadatas,
            documents,
        } = collection_entries;

        let json_body = json!({
            "ids": ids,
            "embeddings": embeddings,
            "metadatas": metadatas,
            "documents": documents,
        });

        let path = format!("/collections/{}/update", self.id);
        let response = self.api.post(&path, Some(json_body)).await?;

        match response.error_for_status() {
            Ok(_) => Ok(()),
            Err(e) => {
                Err(e.into())
            }
        }
    }
```

Let me know what you think.

danwritecode commented:
I also updated the openai embedding module to use reqwest and made the requisite changes there.

Anush008 commented:
Hey @danwritecode

Could you please replace the failing 
`CakeCrusher/chroma@` action with

```yaml
    services:
      chroma:
        image: chromadb/chroma:0.5.0
        ports:
          - 8000:8000
```
?

danwritecode commented:
@Anush008 the test action is updated. I think the only thing left to address now is the Bert module but I'm not sure how to test it or if anything even needs to change with it. 

Lmk your thoughts.

Anush008 commented on Cargo.toml:
```diff
@@ -30,6 +30,11 @@ version = "0.21"
+[dependencies]
+async-trait = "0.1.83"
+reqwest = { version = "0.11", features = ["json"] }
+tokio = { version = "1.0", features = ["full"] }
```
Shouldn't the rt and macros features suffice?

Anush008 replied:
Also, should be a dev dependency?

Anush008 replied:
Hey @danwritecode.
Where is Tokio used outside of tests?

danwritecode replied:
you're right, just updated it.

Anush008 commented:
> Bert module but I'm not sure how to test it or if anything even needs to change with it.
Should be fine I think.

Could you please update the `README.md` and `lib.rs`'s docstring with the new usage please?

danwritecode commented:
> > Bert module but I'm not sure how to test it or if anything even needs to change with it.
> > Should be fine I think.
> 
> Could you please update the `README.md` and `lib.rs`'s docstring with the new usage please?

Updated the README...the lib.rs docstring was already updated with all of the latest.

Also no worries on contributing! I was either going to build my own client or contribute here, so I'm happy to help.

danwritecode commented:
btw this test is still failing, as I noted above I don't think the test is valid anymore. I can remove it but wanted your thoughts. 

```
test v2::collection::tests::test_get_from_collection ... FAILED
```

Anush008 commented:
Fine by me.

danwritecode commented:
> Fine by me.

Okay, it's been removed. I think we're good to go!

Anush008 commented:
Hey @danwritecode.
I'll fix the release CI soon.
